### PR TITLE
Support shared memory in CuPy JIT

### PR DESCRIPTION
**Pull request by asi1024**
Part of #4290

### CODE REVIEW

File: tests/cupyx_tests/jit_tests/test_raw.py [hunk 1]
```diff
@@ -84,6 +84,25 @@ def f(x, y, buf):
...
+        def f(x, y):
+            tid = jit.threadIdx.x
+            ntid = jit.blockDim.x
+            bid = jit.blockIdx.x
+            i = tid + bid * ntid
+
+            smem = jit.shared_malloc(numpy.int32, 32)
+            smem[tid] = x[i]
+            jit.syncthreads()
+            y[i] = smem[ntid - tid - 1]
```
Comment (asi1024) [inline tests/cupyx_tests/jit_tests/test_raw.py:98]: note: The following code is generated.
```cpp
extern "C" __global__ void f(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y) {
  unsigned int tid;
  unsigned int ntid;
  unsigned int bid;
  unsigned int i;
  __shared__ int _smem1[32];
  int* smem;
  tid = threadIdx.x;
  ntid = blockDim.x;
  bid = blockIdx.x;
  i = (tid + (bid * ntid));
  smem = _smem1;
  smem[tid] = x[i];
  __syncthreads();
  y[i] = smem[((long long)((ntid - tid)) - 1ll)];
}
```

Comment (emcastillo) [reply tests/cupyx_tests/jit_tests/test_raw.py:98]: wow, this is pretty awesome!

**Comment by leofang** [comment]
This is very cool! Quick question: Currently this PR only supports static shared memory, not dynamic one (size determined at kernel launch time), right?

**Comment by asi1024** [comment]
@leofang Yes, but we have a plan for the support of dynamic shared memory by `jit.shared_memory(size=None)`

**Comment by asi1024** [comment]
@leofang Added dynamic shared memory in this PR! 1076b31

**Review by emcastillo — APPROVED** [review]

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Awesome!!! Thanks @asi1024!

**Comment by emcastillo** [comment]
Jenkins completed fine, but it was not reporting, manually merging

**Pull request merged**